spark/Debugging-extended-Spark-History-Server.md

```diff
@@ -0,0 +1,202 @@
+3. Expand the group **Custom spark2-defaults**.
+4. Click **Add Property**, add **spark.ui.enhancement.enabled=false**, save.
+5. The property sets to **false** now.
+6. Click **Save** to save the configuration.
+
+    ![feature turns off](./media/apache-spark-history-server/sparkui-turn-off.png)
+
+7. Click **Spark2** in left panel, under **Summary** tab, click **Spark2 History Server**.
+
+    ![restart server1](./media/apache-spark-history-server/sparkui-restart-1.png) 
+
+8. Restart history server by clicking **Restart** of **Spark2 History Server**.
+
+    ![restart server2](./media/apache-spark-history-server/sparkui-restart-2.png)  
+
+9. Refresh the Spark history server web UI, it will be reverted to community version.
+
+
+## 3. Collect history server events for diagnosis 
+
+If you run into history server error, follow the steps below to collect the Spark event:
+1. Download event by clicking **Download** in history server web UI.
+
+    ![download event](./media/apache-spark-history-server/sparkui-download-event.png)
+
+2. Click **Provide us feedback** from data/graph tab.
+
+    ![graph feedback](./media/apache-spark-history-server/sparkui-graph-feedback.png)
+
+3. Provide the title and description of error, drag the zip file to the edit field, then click **Submit new issue**.
+
+    ![file issue](./media/apache-spark-history-server/sparkui-file-issue.png)
+
+
+## 4. Upload history server log using script action
+
+If you run into history server error, use the script below which will upload the history server log to the blob storage specified by us (who is working on investigating the history server issues). 
+
+**upload_shs_log.sh**:
+
+   ```bash
+    #!/usr/bin/env bash
+
+    # Copyright (C) Microsoft Corporation. All rights reserved.
+
+    # Arguments:
+    # $1 Blob link with SAS token query string 
+    # $2 Log path
+    # $3 Max log file size in MB
+
+    if [ "$#" -ne 3 ]; then
+        >&2 echo "$@"
+        >&2 echo "Please provide Azure Storage link, log path and max log size."
+        exit 1
+    fi
+
+    blob_link=$1
+    log_path=$2
+    max_log_size=$3
+
+    if ! [ -e "$log_path" ]; then
+        >&2 echo "There is no log path $log_path on this node"
+        exit 0 
+    fi
+
+    tail -c $((1024*1024*max_log_size)) \
+        "$log_path" > /tmp/shs_log_for_trouble_shooting.log
+
+    curl -T /tmp/shs_log_for_trouble_shooting.log \
+        -X PUT \
+        -H "x-ms-date: $(date -u)" \
+        -H "x-ms-blob-type: BlockBlob" \
+        "$blob_link"
+   ```
+
+**Usage**: 
+
+`upload_shs_log.sh "${blob_link}" ${log_path} ${log_max_MB_size}`
+
+**Example**:
+
+`upload_shs_log.sh "https://${account_name}.blob.core.windows.net/${blob_container}/${log_file}{SAS_query_string}" /var/log/spark2/spark-spark-org.apache.spark.deploy.history.HistoryServer-1-{head_node_alias}-spark2.out 100`
+
+For **head_node_alias**, it may be **hn0** or **hn1** for a cluster with two head nodes. Fill in the active head node alias.
```

Comment:
can the script determine this automatically ?

Reply:
After invesigation, there are several issues: 
1. The log folder contains many files and it is better to avoid hard code the file name. One workaround is the script can copy all of the files however the size might be large and it is hard to do top on files. Therefore we'd like to let DRI to pick the right file to upload
2. This script action is ad-hoc script to collect one-time log from single cluster, the ROI of automation is relatively low

Therefore, prefer to keep this script as it is.